### ISSUE TITLE
[luci-interpreter] Support InstanceNorm op with 3D input.

### ISSUE THREAD

**Issue by ragmani** [Samsung/ONE#13700]
## What

Let's figure out it is possible for instance normalization op to have 3D input.

The following model has InstanceNorm op with 3D input by converting [torch.nn.GroupNorm](https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html) 

![image](https://github.com/user-attachments/assets/ce23df6a-f3a6-4f09-832e-ed8e05f48eef)

## Why

`torch.onnx.export` converts `torch.nn.GroupNorm` into a onnx model containing instance normalization with 3D input such as the following model:

![image](https://github.com/user-attachments/assets/e6d3364f-93e2-4923-b97a-986588036877)

We need to support converting the op and validating it with luci-interpreter unless `torch.onnx.export` supports converting it to [GroupNoralization](https://onnx.ai/onnx/operators/onnx__GroupNormalization.html) of onnx.

## Definition

Currently, TFLite doesn't have InstanceNormalization op and keras also doesn't have InstanceNormalization layer. Therefore, we need to define the corresponding op with 3D input, such as in which input dimension the channel is located or what the rank of the beta input shape is. Fortunately, as shown in https://github.com/Samsung/ONE/issues/13700#issuecomment-2309812869, we can deduce how the corresponding op info should be composed through the network of decomposed GroupNormalization layer.
As a result, InstanceNormalization op with 3D input can be defined as follows:
1. 3D input shape : (batch, channel, unit size to be normalized)
2. beta shape : the shape can be broadcasted to (1, channel, unit size to be normalized)

---

draft #13717

/cc @seanshpark

**Comment by seanshpark** [comment]
> ... but an error occurred.

try clean build

**Comment by ragmani** [comment]
> try clean build

Build success :)

**Comment by seanshpark** [comment]
To check with value test, can you please add new recipe with 3D, and test to `compiler/luci-value-py-test/test.lst` ?

**Comment by ragmani** [comment]
How do I add the recipe?

**Comment by ragmani** [comment]
Unfortunately, cannot use `luci-value-py-test` to simply add tests because tflite does not support InstanceNorm op.
Let's use `luci-pass-value-py-test` instead of `luci-value-py-test`.

**Comment by ragmani** [comment]
An error occurred after adding the test for fuse_instnorm.

![image](https://github.com/user-attachments/assets/882d2878-cc94-468d-b4ec-618c3b4b1cb1)


```bash
[ 58%] Built target luci_interpreter_memory_manager_test
[ 58%] Generate Net_InstanceNorm_008.recipe
[ 58%] Generate Net_InstanceNorm_008.rule
[ 58%] Generate Net_InstanceNorm_008 nnpackage test directory
[ 58%] Generate Net_InstanceNorm_008.tflite
[ 58%] Generate Net_InstanceNorm_008.circle
[ 58%] Generate Net_InstanceNorm_008.opt.circle
ERROR: Optimized graph is invalid
...
```

I guess InstanceNorm op fusion is not supported for 3D I/O.

**Comment by ragmani** [comment]
![image](https://github.com/user-attachments/assets/ac2acdf9-c226-48b0-9966-81d88365d79b)
![image](https://github.com/user-attachments/assets/a3a36d47-93f1-499a-a1db-6ca03bc2faca)

`beta` input is a 1D vector in instancenorm op. Nevertheless, is it okay to create a recipe assuming that the constant value of Sub op has been fused with reshape op(32 -> 1 x 32 x 1) like below?
![image](https://github.com/user-attachments/assets/6744f5e0-84c8-41ce-977b-d27671523107)

**Comment by ragmani** [comment]
> beta input is a 1D vector in instancenorm op. Nevertheless, is it okay to create a recipe assuming that the constant value of Sub op has been fused with reshape op(32 -> 1 x 32 x 1) like below?

There is no InstanceNormalization layer in keras, but we can create a model with InstanceNormalization layer by using [GroupNormalization](tf.keras.layers.GroupNormalization).

![image](https://github.com/user-attachments/assets/3420632c-b47b-40bf-b2bd-07f1c50984e4)

The constant value of Sub op is 3D, not 1D in the tflite model.

**Comment by seanshpark** [comment]
>  is it okay to create a recipe assuming that the constant value of Sub op has been fused with reshape op(32 -> 1 x 32 x 1) like below?

I'm not sure the meaning of `fused with reshape` but `{1x32x1} = Sub({1x32x1}, {1x32x1})` looks ok.

**Comment by ragmani** [comment]
>>  is it okay to create a recipe assuming that the constant value of Sub op has been fused with reshape op(32 -> 1 x 32 x 1) like below?

> I'm not sure the meaning of fused with reshape but {1x32x1} = Sub({1x32x1}, {1x32x1}) looks ok.

"fused with reshape op" means constant folding of the formula `Reshape(input(32), shape(1, 32, 1))` to `input(1, 32, 1)`

**Comment by ragmani** [comment]
@seanshpark 
I appended the definition of InstanceNorm op with 3D input in circle. Please let me know if there are anything strange or definitions to be add.

**Comment by seanshpark** [comment]
> I appended the definition of InstanceNorm op ...

looks good to me :)

**Comment by ragmani** [comment]
Done.
